lib/JSON/Validator.pm

```diff
@@ -567,15 +569,20 @@ sub _validate {
     if ref $schema eq 'HASH' and $schema->{'$ref'};
   return $schema ? () : E $path, [not => 'not'] if is_type $schema, 'BOOL';
 
-  my $seen_addr = join ':', refaddr($schema),
-    (ref $data ? refaddr $data : ++$self->{seen}{scalar});
+  my @errors;
+
+  if ( $self->recursive_data_protection ) {
+    my $seen_addr = join ':', refaddr($schema),
+      (ref $data ? refaddr $data : ++$self->{seen}{scalar});
+
+    # Avoid recursion
+    if ($self->{seen}{$seen_addr}) {
+      return @{$self->{seen}{$seen_addr}};
+    }
```

Comment:
Please put this block on one line to save LOC.

Reply:
Moved to a single line.